Index: src/engines/SoGate.cpp
===================================================================
RCS file: /export/cvsroot/Coin-2/src/engines/SoGate.cpp,v
retrieving revision 1.1.1.1
diff -u -r1.1.1.1 SoGate.cpp
--- src/engines/SoGate.cpp	27 Feb 2003 15:37:31 -0000	1.1.1.1
+++ src/engines/SoGate.cpp	22 Sep 2003 14:56:32 -0000
@@ -173,6 +173,7 @@
   this->dynamicoutput = new SoEngineOutputData(SoGate::outputdata);
   this->dynamicoutput->addOutput(this, "output", this->output, inputfieldtype);
   this->output->setContainer(this);
+  this->output->enable(FALSE);
 }
 
 /*!
@@ -191,7 +192,9 @@
 SoGate::evaluate(void)
 {
   // Force update of slave fields.
-  this->output->enable(TRUE);
+  // FIXME : why should this be necessary ? it only promotes unwanted evaluation
+  // at initialisation. Gerhard Reitmayr
+  // this->output->enable(TRUE);
 
   // FIXME: this is an amazingly ineffective way of copying the
   // values. 20000920 mortene.
